File: neighbors2.ml

type p = {
  i: int;
  j: int;
}

let mk_p i j = {i; j}

let equal_p p1 p2 = p1.i = p2.i && p1.j = p2.j

let hash_p p = p.i + p.j * 4000 

let iter_neighbors f {i; j} =
  f (mk_p (i-1) j);
  f (mk_p (i+1) j);
  f (mk_p i (j-1));
  f (mk_p i (j+1));
  ()

module PTbl = Hashtbl.Make(struct
    type t = p
    let equal = equal_p
    let hash = hash_p
  end)

type hashset = unit PTbl.t

let rec nth_loop n (s1:hashset) (s2:hashset) =
  match n with
    | 0 -> s1
    | _ ->
      let s0 = PTbl.create 32 in
      let add p =
        if not (PTbl.mem s1 p || PTbl.mem s2 p)
        then PTbl.replace s0 p ()
      in
      PTbl.iter
        (fun p' () -> iter_neighbors add p') s1;
      nth_loop (n-1) s0 s1

let nth n p =
  let s1 = PTbl.create 32 in
  PTbl.add s1 p ();
  let s2 = PTbl.create 32 in
  nth_loop n s1 s2

let () =
  Printf.printf "result is %d\n" (nth 2000 (mk_p 0 0) |> PTbl.length)

File: neighbors2.fs
open System
open System.Collections.Generic

type [<Struct>] P =
  val i : int
  val j : int
  new(i, j) = {i=i; j=j}

let cmp =
  { new System.Collections.Generic.IEqualityComparer<P> with
      member __.Equals(this, that) =
        this.i = that.i && this.j = that.j
      
      member __.GetHashCode this = this.i + 4000 * this.j }

let inline iterNeighbors f (p: P) =
  let i, j = p.i, p.j
  f(P(i-1, j))
  f(P(i+1, j))
  f(P(i, j-1))
  f(P(i, j+1))

let rec nthLoop n (s1: HashSet<_>) (s2: HashSet<_>) =
  match n with
  | 0 -> s1
  | n ->
      let s0 = HashSet(cmp)
      let add p =
        if not(s1.Contains p || s2.Contains p) then
          ignore(s0.Add p)
      Seq.iter (fun p -> iterNeighbors add p) s1
      nthLoop (n-1) s0 s1

let nth n p =
  nthLoop n (HashSet([p], cmp)) (HashSet(cmp))

Printf.printf "result is %d\n" (nth 2000 (P(0, 0))).Count
